File: samples/dashDBRuby/dashdbsample.rb
require 'rubygems'
if RUBY_PLATFORM =~ /mingw/
  require 'mswin32/ibm_db'
else
  require 'ibm_db'
end
require 'sinatra'
require 'json'

# @Author IBM
# @version 0.1

IS_CF=true

get '/' do

  if IS_CF
      jsondb_db = JSON.parse(ENV['VCAP_SERVICES'])["dashDB"]
	  halt 500, "No dashDB service bound. Please bind a dashDB Service instance" if jsondb_db.nil? 
	  credentials = jsondb_db[0]["credentials"]
      host = credentials["host"]
      username = credentials["username"]
      password = credentials["password"]
      database = credentials["db"]
      port = credentials["port"]
      dsn = "DRIVER={IBM DB2 ODBC DRIVER};DATABASE="+database+";HOSTNAME="+host+";PORT="+port.to_s()+";PROTOCOL=TCPIP;UID="+username+";PWD="+password+";"
    else
	  halt 500, "no vcap services"
    end

out = String.new
total = String.new

if conn = IBM_DB.connect(dsn, '', '')
  
  # run a sample select query on the catalog tables
  sql = "SELECT FIRST_NAME, LAST_NAME from GOSALESHR.employee FETCH FIRST 10 ROWS ONLY"
  
  begin
    if stmt = IBM_DB.exec(conn, sql)
      # iterate through the resultset
	  out = "10 rows from table GOSALESHR.employee <table border=\"1\"><tr><td>Firstname</td><td>LastName</td>"
	  total = total + out + "<BR>\n"
      while row = IBM_DB.fetch_assoc(stmt)
        out = "<tr><td>#{row['FIRST_NAME']}</td><td>#{row['LAST_NAME']}</td>"
        total = total + out + "<BR>"
      end
	  
	  out = "</table>"
	  total = total + out + "<BR>\n"
      # free the resources associated with the result set
      IBM_DB.free_result(stmt)
    else
      halt 500 , "Statement execution failed: #{IBM_DB.stmt_errormsg}"
    end
  ensure
    # Cleanup close the database connection
    
    IBM_DB.exec(conn, sql)
    
    IBM_DB.close(conn)
  end
else
  halt 500 , "Connection failed: #{IBM_DB.conn_errormsg}"
  
end
total
end

File: samples/dashDBJava/dashDBSample/WEB-INF/classes/com/ibm/dashdb/sample/DashDBSample.java
package com.ibm.dashdb.sample;
import java.io.IOException;
import java.io.PrintWriter;
import java.net.InetAddress;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.Set;

import javax.servlet.ServletException;
import javax.servlet.annotation.WebServlet;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import com.ibm.nosql.json.api.BasicDBList;
import com.ibm.nosql.json.api.BasicDBObject;
import com.ibm.nosql.json.util.JSON;

// @author Prrsahoo, IBM 2014
// @version 0.2

/**
 * Servlet implementation class BLUDWSample
 */
@WebServlet("/DashDBSample")
public class DashDBSample extends HttpServlet {
	private static final long serialVersionUID = 1L;

	/**
	 * @see HttpServlet#HttpServlet()
	 */
	public DashDBSample() {
		super();
	}

	/**
	 * @see HttpServlet#doGet(HttpServletRequest request, HttpServletResponse
	 *      response)
	 */

	protected void doGet(HttpServletRequest request,
			HttpServletResponse response) throws ServletException, IOException {
		response.setContentType("text/plain");
		response.setStatus(200);
		PrintWriter writer = response.getWriter();
		writer.println("IBM dashDB, Java Demo Application using DB2 drivers");
		writer.println("Servlet: " + this.getClass().getName());
		writer.println();
		writer.println("VCAP Host:" + System.getenv("VCAP_APP_HOST") + ":"
				+ System.getenv("VCAP_APP_PORT"));
		writer.println("Host IP:" + InetAddress.getLocalHost().getHostAddress());

		BasicDBObject bludb = null;

		// VCAP_SERVICES is a system environment variable
		// Parse it to obtain the for DB2 connection info
		String VCAP_SERVICES = System.getenv("VCAP_SERVICES");
		writer.println("VCAP_SERVICES content: " + VCAP_SERVICES);

		if (VCAP_SERVICES != null) {
			// parse the VCAP JSON structure
			BasicDBObject obj = (BasicDBObject) JSON.parse(VCAP_SERVICES);
			String thekey = null;
			Set<String> keys = obj.keySet();
			writer.println("Searching through VCAP keys");
			// Look for the VCAP key that holds the dashDB information
			for (String eachkey : keys) {
				writer.println("Key is: " + eachkey);
				// The old name for this service was AnalyticsWarehouse
				if (eachkey.contains("dashDB")) {
					thekey = eachkey;
				}
			}
			if (thekey == null) {
				writer.println("Cannot find any dashDB service in the VCAP; exiting");
				return;
			}

			BasicDBList list = (BasicDBList) obj.get(thekey);
			bludb = (BasicDBObject) list.get("0");
			writer.println("Service found: " + bludb.get("name"));
			bludb = (BasicDBObject) bludb.get("credentials");
			
		} else {
			writer.println("VCAP_SERVICES is null");
			return;
		}

		writer.println();
		writer.println("full database uri: " + bludb.get("uri"));
	
		Connection con = null;

		try {
			writer.println();
			writer.println("Loading the Database Driver");
			Class.forName("com.ibm.db2.jcc.DB2Driver");
			// Connect to the database
			writer.println("Connecting to the database");
			writer.println();
			
			String jdbcurl =  (String) bludb.get("jdbcurl");
			String user = (String) bludb.get("username");
		    String password = (String) bludb.get("password");
			
			con = DriverManager.getConnection(jdbcurl, user,password);
			// or if the username & password are in the url:
			// con = DriverManager.getConnection(databaseUrl);
			// Commit changes manually
			con.setAutoCommit(false);
		} catch (SQLException e) {
			writer.println("Error connecting to database");
			writer.println("SQL Exception: " + e);
			return;
		} catch (ClassNotFoundException e) {
			writer.println("Error loading driver");
			writer.println("Error: " + e);
			return;
		}

		Statement stmt = null;
		String sqlStatement = "";
				
		// Execute some SQL statements on the database 
		try {
			stmt = con.createStatement();
			sqlStatement = "SELECT FIRST_NAME, LAST_NAME from GOSALESHR.employee FETCH FIRST 10 ROWS ONLY";
			ResultSet rs = stmt.executeQuery(sqlStatement);
			writer.println("Executing: " + sqlStatement);

			writer.println("FIRST_NAME \t\t LAST_NAME");
			// Process the result set
			while (rs.next()) {
				writer.println(rs.getString(1) + "\t\t" +  rs.getString(2));
			}
			// Close the ResultSet
			rs.close();

		} catch (SQLException e) {
			writer.println("Error executing:" + sqlStatement);
			writer.println("SQL Exception: " + e);
		}

		// Close everything off
		try {
			// Close the Statement
			stmt.close();
			// Connection must be on a unit-of-work boundary to allow close
			con.commit();
			// Close the connection
			con.close();
			writer.println("Finished");

		} catch (SQLException e) {
			writer.println("Error closing things off");
			writer.println("SQL Exception: " + e);
		}

		writer.close();
	}

	/**
	 * @see HttpServlet#doPost(HttpServletRequest request, HttpServletResponse
	 *      response)
	 */
	protected void doPost(HttpServletRequest request,
			HttpServletResponse response) throws ServletException, IOException {
	}

}
